spec.html

```diff
@@ -26403,6 +26493,170 @@ <h1>Cyclic Module Records</h1>
+        <emu-table id="table-moduleloadstate-record-fields" caption="ModuleLoadState Record Fields">
+          <table>
+            <tr>
+              <th>
+                Field Name
+              </th>
+              <th>
+                Value Type
+              </th>
+              <th>
+                Meaning
+              </th>
+            </tr>
+            <tr>
+              <td>
+                [[Action]]
+              </td>
+              <td>
+                ~graph-loading~ or ~dynamic-import~
+              </td>
+              <td>
+                The action that caused the call to HostLoadImportedModule. It is ~graph-loading~ when loading the dependencies of a module; it is ~dynamic-import~ when loading a module that was requested by an `import()` call.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[PromiseCapability]]
+              </td>
+              <td>
+                a PromiseCapability Record
+              </td>
+              <td>
+                The promise to resolve when the loading process finishes.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[IsLoading]]
+              </td>
+              <td>
+                a boolean
+              </td>
+              <td>
+                This field is only used if [[Action]] is ~graph-loading~. It is true if the loading process has not finished yet, neither successfully nor with an error.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[PendingModules]]
+              </td>
+              <td>
+                a non-negative integer
+              </td>
+              <td>
+                This field is only used if [[Action]] is ~graph-loading~. It tracks the number of pending HostLoadImportedModule calls.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[Visited]]
+              </td>
+              <td>
+                a List of Cyclic Module Records
+              </td>
+              <td>
+                This field is only used if [[Action]] is ~graph-loading~. It is a list of the Cyclic Module Records that have been already loaded by the current loading process, to avoid infinite loops with circular dependencies.
+              </td>
+            </tr>
+            <tr>
+              <td>
+                [[HostDefined]]
+              </td>
+              <td>
+                anything (default value is *undefined*)
+              </td>
+              <td>
+                It contains host-defined data to pass from the LoadRequestedModules caller to HostLoadImportedModule.
+              </td>
+            </tr>
+          </table>
+        </emu-table>
+
+        <emu-clause id="sec-LoadRequestedModules" type="concrete method">
+          <h1>LoadRequestedModules ( [ _hostDefined_ ] ): a Promise object</h1>
```

Comment:
LoadRequestedModules is part of the "public API", so I prefer to keep it this way. I don't think that this spec optimization matters in practice.

Reply:
What do you mean when you say "public API" here?

I do think it is quite a wart to ignore fields, then recreate the same record twice in spec text.

Reply:
It's expected to be called by hosts

Reply:
Right, but within an individual spec load operation, we shouldn't need to overload the internal state while also having two separate records representing that - we should pick one or the other approach.

If you don't want to add an optional record to the "public API", we could instead separate:

* `LoadRequestedModules` -> `1. Let _state_ be a new record with fields... 2. LoadRequestedModulesInternal(_state_)
* Dynamic import would then directly call `LoadRequestedModulesInternal(_state_).

So I don't think the public / private argument causes any issue here?

When you import a module reflection, that's a branch in the graph loader on this edge property of the node, just like it is for dynamic import, and which we will have to write, just like we will have to write branching in dynamic import.

The issue seems to me more to do with the fact that dynamic import has to go through an _initial phase_, before it can start the graph loading algorithm, where that initial phase is defined by the host hook which needs to call back with some state tracking. An alternative might be to use another type of state record to represent this instead of overloading?

Please let me know if I'm missing any further congruences here though.

Reply:
I implemented your suggestion at https://github.com/nicolo-ribaudo/ecma262/pull/3, however I find it way less clear than the current status. It feels like an optimizations that engines might do to avoid instantiating two records, but which isn't necessary at the spec level.

I'll leave the commit there, and defer to the editors the decision whether to do it or not.